SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -425,32 +426,35 @@ - (void)sendMessage:(TSOutgoingMessage *)message
             success:(void (^)())successHandler
             failure:(void (^)(NSError *error))failureHandler
 {
-    [self sendMessage:message transaction:nil success:successHandler failure:failureHandler];
-}
-
-- (void)sendMessage:(TSOutgoingMessage *)message
-        transaction:(YapDatabaseReadWriteTransaction *_Nullable)transaction
-            success:(void (^)())successHandler
-            failure:(void (^)(NSError *error))failureHandler
-{
     OWSAssert(message);
 
-    if (transaction) {
-        [message updateWithMessageState:TSOutgoingMessageStateAttemptingOut transaction:transaction];
-    } else {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        // This method will use a read/write transaction. This transaction
+        // will block until any open read/write transactions are complete.
+        //
+        // That's key - we don't want to send any messages in response
+        // to an incoming message until processing of that batch of messages
+        // is complete.
```

Comment:
This comment is important.

Reply:
So we're using the side effect of yap's one-write-transaction-per-db as a semaphor here?

I can imagine, generally speaking, the problems this intends to address, but do you have a specific example? This comment is more "what and how" than "why".

Reply:
Right. I've elaborated the comment!